rust/lance-encoding/src/statistics.rs

```diff
@@ -0,0 +1,1026 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The Lance Authors
+
+//! Data layouts to represent encoded data in a sub-Arrow format
+//!
+//! These [`DataBlock`] structures represent physical layouts.  They fill a gap somewhere
+//! between [`arrow_data::data::ArrayData`] (which, as a collection of buffers, is too
+//! generic because it doesn't give us enough information about what those buffers represent)
+//! and [`arrow_array::array::Array`] (which is too specific, because it cares about the
+//! logical data type).
+//!
+//! In addition, the layouts represented here are slightly stricter than Arrow's layout rules.
+//! For example, offset buffers MUST start with 0.  These additional restrictions impose a
+//! slight penalty on encode (to normalize arrow data) but make the development of encoders
+//! and decoders easier (since they can rely on a normalized representation)
+
+use std::{fmt, sync::Arc};
+
+use arrow_array::{Array, UInt64Array};
+
+use crate::data::{
+    DataBlock, DictionaryDataBlock, FixedWidthDataBlock, OpaqueBlock, StructDataBlock, VariableWidthBlock
+};
+
+#[derive(Clone, Copy, PartialEq, Eq, Hash)]
+pub enum Stat {
+    BitWidth,
+    DataSize,
+    Cardinality,
+    FixedSize,
+    NullCount,
+}
+
+impl fmt::Debug for Stat {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::BitWidth => write!(f, "BitWidth"),
+            Self::DataSize => write!(f, "DataSize"),
+            Self::Cardinality => write!(f, "Cardinality"),
+            Self::FixedSize => write!(f, "FixedSize"),
+            Self::NullCount => write!(f, "NullCount"),
+        }
+    }
+}
+
+impl fmt::Display for Stat {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+pub trait GetStat {
+    fn get_stat(&mut self, stat: Stat) -> Option<Arc<dyn Array>>;
+}
+
+impl GetStat for DataBlock {
+    fn get_stat(&mut self, stat: Stat) -> Option<Arc<dyn Array>> {
+        match self {
+            Self::AllNull(_) => {
+                //  the statistics is not calculated here as this enum is going to deprecated soon anyway
+                None
+            }
+            Self::Nullable(_) => {
+                //  the statistics is not calculated here as this enum is going to deprecated soon anyway
+                None
+            }
+            Self::FixedWidth(data_block) => data_block.get_stat(stat),
+            Self::FixedSizeList(data_block) => data_block.child.get_stat(stat),
```

Comment:
Cardinality and bit-width of a fixed-size-list will technically be different than the cardinality and bit-width of the child.  Maybe just leave this as a `todo!()` or `None` for now.